_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+    "active": true
+}
+```
+
+What response do you expect? If your code is working correctly, the Todo app backend will give a response with status 201 Created, and the following body: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:schemas:core:2.0:User"
+    ],
+    "locale": "en-US",
+    "groups": [],
+    "meta": {
+        "resourceType": "User"
+    },
+    "id": "5",
+    "userName": "test.user@portal.example",
+    "name": {
+        "givenName": "Test",
+        "familyName": "User"
+    },
+    "emails": [
+        {
+            "primary": true,
+            "value": "test.user@portal.example",
+            "type": "work"
+        }
+    ],
+    "displayName": "Test User",
+    "externalId": "00ujl29u0le5T6Aj10h7",
+    "active": true
+}
+```
+
+If you send the same POST request again, re-creating an existing user, what should happen? Try it, and you'll get a 409 Conflict response, with a body like this: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:api:messages:2.0:Error"
+    ],
+    "detail": "User already exists in the database: 5",
+    "status": 409
+}
+```
+
+You can repeat this testing process for each SCIM route that you implement! 
+
+### GET /Users
+According to [the SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.4), you may return users paginated. [Okta requests users paginated]( https://developer.okta.com/docs/reference/scim/scim-20/#retrieve-users) with a Start Index set to 1 and a count set to 100 per page. 
+
+In addition, filtering by username is an OPTIONAL parameter for SCIM service providers refer to this section of [the SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2). However, Okta requires username filtering support as it checks to see if the user exists by username first before creating them. Ref: [Okta Docs](https://developer.okta.com/docs/reference/scim/scim-20/#create-users)
+
+To fulfill these requirements, the sample app's backend can handle GET requests to the `/Users` endpoint in `scim.ts` like this: 
+
+```
+// Retrieve Users
+// GET /scim/v2/Users
+// RFC Notes on Retrieving Users: https://www.rfc-editor.org/rfc/rfc7644#section-3.4.1
+scimRoute.get('/Users', passport.authenticate('bearer'), async (req, res) => {
+
+
+    // console.log('GET: /Users');
+    // console.log(req.query.filter)
+    // RFC Notes on Pagination: https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.4
+    const DEFAULT_START_INDEX = '1';
+    const DEFAULT_RECORD_LIMIT = '100';
+    let startIndex = parseInt(req.query.startIndex as string ?? DEFAULT_START_INDEX);
+    startIndex--; // Prisma starts its indexing at 0, while Okta send starting index at 1, this is to make sure we start at 0
+    const recordLimit = parseInt(req.query.recordLimit as string ?? DEFAULT_RECORD_LIMIT);
+    let filterQuery: string | null = req.query.filter as string ?? null;
+    let filterParams: string[] = [];
+    let email = null;
+    if (!!filterQuery) {
```

Comment:
is the double `!!` really right here? is this a normal JS thing because it feels tricky and looks wrong. I'd suggest either rewriting it to be less tricky or explaining what's going on.

Reply:
standard typescript to cast it to a boolean